### PR TITLE
C++: Use an IPA type rather than negative indexes for argument/parameter matching in data flow

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

**Review by MathiasVP — COMMENTED** [review]
Only a couple of minor comments. I think we should run a DCA for this as well before we merge it.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 1]
```diff
@@ -490,19 +491,52 @@ class ExprNode extends InstructionNode {
...
-  //  1 -> -4
-  // ...
-  result = -3 - index
+/** A parameter position represented by an integer. */
+class ParameterPosition = Position;
+
+/** An argument position represented by an integer. */
+class ArgumentPosition = Position;
+
+class Position extends TPosition {
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:500]: Good point, but I think I can move it to DataFlowPrivate, actually.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 2]
```diff
@@ -11,6 +11,7 @@ private import semmle.code.cpp.ir.IR
 private import semmle.code.cpp.controlflow.IRGuards
 private import semmle.code.cpp.models.interfaces.DataFlow
 private import DataFlowPrivate
+private import DataFlowDispatch
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:14]: I think I added that before moving the `ArgumentPosition` definitions out of `DataFlowDispatch`. It's not needed any more.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 3]
```diff
@@ -490,19 +491,52 @@ class ExprNode extends InstructionNode {
...
+    index != -1 and
+    result = index.toString()
+  }
+
+  int getIndex() { result = index }
 }
 
+newtype TPosition =
+  TDirectPosition(int index) { exists(any(CallInstruction c).getArgument(index))} or
+  TIndirectionPosition(int index) { exists(any(CallInstruction c).getArgument(index)) }
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:538]: Probably. I'll just check `ReadSideEffectInstruction::getIndex()`, though.

**Comment by rdmarsh2** [comment]
DCA run: https://github.com/github/codeql-dca-main/tree/data/rdmarsh2/dataflow-ipa-pa__Differences__code-scanning__1/reports

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**